pkg/runtime/framework/plugins/jobset/jobset.go

```diff
@@ -106,55 +142,32 @@ func (j *JobSet) Build(ctx context.Context, info *runtime.Info, trainJob *traine
 		return nil, nil
 	}
 
-	// Get the runtime as unstructured from the TrainJob ref
-	runtimeJobTemplate := &unstructured.Unstructured{}
-	runtimeJobTemplate.SetAPIVersion(trainer.GroupVersion.String())
-
-	if kind := trainJob.Spec.RuntimeRef.Kind; kind != nil {
-		runtimeJobTemplate.SetKind(*kind)
-	} else {
-		runtimeJobTemplate.SetKind(trainer.ClusterTrainingRuntimeKind)
-	}
-
-	key := client.ObjectKey{Name: trainJob.Spec.RuntimeRef.Name}
-	if runtimeJobTemplate.GetKind() == trainer.TrainingRuntimeKind {
-		key.Namespace = trainJob.Namespace
-	}
-
-	if err := j.client.Get(ctx, key, runtimeJobTemplate); err != nil {
-		return nil, err
-	}
-
-	// Populate the JobSet template spec apply configuration
-	jobSetTemplateSpec := &jobsetv1alpha2ac.JobSetSpecApplyConfiguration{}
-	if jobSetSpec, ok, err := unstructured.NestedFieldCopy(runtimeJobTemplate.Object, "spec", "template", "spec"); err != nil {
-		return nil, err
-	} else if !ok {
-		return nil, fmt.Errorf("trainJob runtime %s does not have a spec.template.spec field", trainJob.Spec.RuntimeRef.Name)
-	} else {
-		if raw, err := json.Marshal(jobSetSpec); err != nil {
-			return nil, err
-		} else if err := json.Unmarshal(raw, jobSetTemplateSpec); err != nil {
-			return nil, err
-		}
-	}
-
 	// Init the JobSet apply configuration from the runtime template spec
 	jobSetBuilder := NewBuilder(jobsetv1alpha2ac.JobSet(trainJob.Name, trainJob.Namespace).
 		WithLabels(maps.Clone(info.Labels)).
 		WithAnnotations(maps.Clone(info.Annotations)).
-		WithSpec(jobSetTemplateSpec))
+		WithSpec(jobSetSpec))
```

Comment:
Should we apply `jobSetSpec` to the JobSet spec after we run the JobSet builder ?
Since Plugins always contains the final values for JobSet even if TrainJob overrides it.
For example, if webhook is disabled and if user accidentally configures `PET_NNODES` envs in the TrainJob, we have to override it according to the Torch plugin.

Reply:
Yeah, that sounds reasonable. Ideally, we want to propagate information by `Info.TemplateSpec.PodSet -> TrainJob -> TrainingRuntime`. In this case , input TrainingRuntime will be stored in `Info.TemplateSpec.ObjApply` as immutable, and each plugin update only `Info.TemplateSpec.PodSet`.

But, here problem is `Info.Trainer`. So, I would like to do more refactor in the follow-up PR.

Reply:
Sounds good.